compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp

```diff
@@ -679,6 +679,79 @@ static Value rewriteCatNegateAndSlice(RewriterBase &rewriter,
   return createCatNegateAndSlice(rewriter, outTensor, source);
 }
 
+// Match an extract slice that has a shape of size N and an output shape of size
+// N-1, where the extraction happened along only the innermost dimension.
+//
+// e.g.
+// %extracted_slice = tensor.extract_slice
+//   %1[0, 0, 0, 1] [1024, 7, 7, 1] [1, 1, 1, 1]
+//   : tensor<1024x7x7x2xf32> to tensor<1024x7x7xf32>
+// Rewrite by adding a transpose that makes dim N the innermost and slice along
+// that dim
+static void matchAndRewriteInnerExtractSlice(RewriterBase &rewriter,
+                                             tensor::ExtractSliceOp sliceOp) {
+
+  RankedTensorType oldInputType = sliceOp.getSource().getType();
+  SmallVector<int64_t> oldInputShape(oldInputType.getShape());
+  ArrayRef<int64_t> oldOutputShape = sliceOp.getResult().getType().getShape();
+
+  if (llvm::is_contained(oldInputShape, ShapedType::kDynamic) ||
+      llvm::is_contained(oldOutputShape, ShapedType::kDynamic)) {
+    return;
+  }
+
+  SmallVector<int64_t> offsets(sliceOp.getStaticOffsets());
+  SmallVector<int64_t> sizes(sliceOp.getStaticSizes());
+  SmallVector<int64_t> strides(sliceOp.getStaticStrides());
+
+  // Match extract extraction of full tensor of shape n-1
+  if (oldInputShape.size() < 2 ||
+      oldInputShape.size() != oldOutputShape.size() + 1 || sizes.back() != 1) {
+    return;
+  }
+
+  // All dims of input match output except for extracted innermost dim
+  for (auto [in, out] : llvm::zip(oldInputShape, oldOutputShape)) {
+    if (in != out) {
+      return;
+    }
+  }
+
+  OpBuilder::InsertionGuard guard(rewriter);
```

Comment:
I haven't used an insertion guard before. I assume it allows you to set the insertion point manually, then when it gets destroyed (i.e., at the end of the function scope), the insertion point resets to where it was before?

Reply:
yes, that's what it does. It's good in cases like this where we aren't using something like the GreedyPatternRewriter to manage insertion points for us.